### ISSUE TITLE
Add option to dry-run Infection

### ISSUE THREAD

**Issue by sanmai** [infection/infection#1150]
When testing Infection on a big project you don't want to actually MT that project. Because it'll take a lot of time, and because you are testing Infection, but not that project.

So far I had to comment out `start()` call inside `ParallelProcessRunner`, and it works. But this interferes with testing. I have to un-comment it back to run PHPUnit, and comment it again to benchmark Infection. 

Another problem here is that the report gets filled with `E`'s. Therefore, some parts of our processing pipeline become ignored and unused. Say, we don't necessarily collect errored mutations for future logging. What if we are hitting a bottleneck if we do?

What I want is to be able to run Infection with `--dry-run[=N]`, where N is an optional random seed, not unlike that of PHPUnit. Then Infection would use a stand-in dummy process that will fail/err/escape pseudo-randomly according to the seed.

**Comment by theofidry** [comment]
I've been thinking the same since we started to profile more regularly: executing the processes seriously skews the results. Generally the more I/O and network we have the less the profiles are reliable.

So I'm totally up for a `--dry-run` option. Maybe the solution is to replace the parallel runner which will consume the provided generator (to trigger its content) but won't fire a process. It may just not dispatch any `MutantProcessWasFinished` and we could skip the whole logging with a `NullLogger` for example.

I think we if are more interesting in the downstream part, e.g. the loggers, we are better off with finding a way to benchmark it without all the initial run

**Comment by sanmai** [comment]
As of writing we're collecting 100% of execution results, even if we can tell for certain we won't need some of them, or all of them, because of disabled logging. This is a definite area of improvement, but with the current state of `--dry-run` it is impossible to see any benefits from changes in the area.

**Comment by theofidry** [comment]
@sanmai I do not understand what is missing here. Could you detail it further?

**Comment by sanmai** [comment]
PR #1183 implemented `--dry-run` by skipping the actual process execution, but the code still:  

- Collects all of execution results                                                                                                                                                                                                                                                            
- Processes all of them through the result pipeline                                                                                                                                                                                                                                             
- Does all kinds of downstream work (event, metrics, etc.)    

Ideally we'd want to be able to profile mutation generation in isolation, without the noise of result collection. And possibly result collection without mutation generation.

```php
enum DryRunMode {                                                                 
    case Off;                    // Normal execution                              
    case SkipExecution;          // Current behavior: skip processes, still collec
    case SkipCollection;         // Skip processes AND skip event dispatch        
}                                                                                 
```

**Comment by theofidry** [comment]
Hmm I think it depends what you want to use it for. If you wish to see what infection is going to do without the most expensive part, then I think the current implementation makes sense (except maybe some reports).

Collecting the initial tests still feels necessary unfortunately, unless you're passing `--only-covered` and giving the coverage report _then_ there is unnecessary work being done. But it's a big of a peculiar loop and I wouldn't expect it to be the default behaviour implied by `--dry-run`.

>Ideally we'd want to be able to profile mutation generation in isolation, without the noise of result collection

I am a bit conflicted here.

_At first_ I was going to suggest to create a dedicated command for that purpose, it seems more appropriate than twisting the run command for something that is far from the typical user journey (for a run command).

But looking at the current benchmark code, you'll inevitably profile bits that are a bit uninteresting.

Ok but actually, nonetheless, maybe it will be sufficient. I will sketch out something and you'll tell me if that fits the bill.

**Comment by sanmai** [comment]
To be fair, it was quite a few years ago, so I struggle to remember what I was intending to do with it all.

**Comment by theofidry** [comment]
Reviewing this issue a bit more freshly.

> What I want is to be able to run Infection with --dry-run[=N], where N is an optional random seed, not unlike that of PHPUnit. Then Infection would use a stand-in dummy process that will fail/err/escape pseudo-randomly according to the seed.

To me this sounds more like a "fake run", i.e. a run where we fake out results, rather than a dry-run.

We also have `--noop` which does execute the processes without applying the mutation.

For profiling we have `make profile_mutation_generator`.

So I think the current dry-run is good enough for the current purpose: executing infection without doing the expansive killing mutants parts. It may seem a bit redundant with `--noop`, but I think it's fine.

So unless you can think of something else (which does not deserve its own issue), this could be closed.

**Comment by sanmai** [comment]
Yes, it is a more like a fake run; where this can be used is to profile Infection's memory usage on real-world projects as we retain various details about mutations up until the very end depending on the outcome. And a predictable seed can show that for a given set of escaped/killed mutations Infection now uses less memory, runs faster, etc, without having to rely on implicit randomness of the actual project on hand, or wait for Infection to do the whole set of work.

**Comment by sanmai** [comment]
I'm going to close this for now, as I don't have the capacity to work on this, and this isn't fair for anyone else to work on this.
